### ISSUE TITLE
Implement type inference for ReduceScatterOp

### ISSUE THREAD

**Issue by burmako** [openxla/stablehlo#864]
To quote the spec, "`type(result) = type(operand)` except `dim(result, scatter_dimension) = dim(operand, scatter_dimension) / dim(process_groups, 1)`". However, at the moment we cannot compute "dim(process_groups, 1)" because it depends on `num_replicas` and `num_partitions`. We'll need to resolve this.

**Comment by burmako** [comment]
Sorry - I haven't provided a rationale for why I unassigned this ticket.

`dim(process_groups, 1)` is something that we cannot compute at the moment. It needs `num_replicas` and `num_partitions`, and that's not something that is reflected in the StableHLO dialect right now. #425 is planning to fix this by adding this information to the StableHLO opset (and the dialect).

In the meanwhile, while this work is blocked on another ticket, I've unassigned this ticket.
